src/pymor/algorithms/krylov.py

```diff
@@ -38,21 +40,38 @@ def arnoldi(A, E, b, r):
         Orthonormal basis for the Krylov subspace as a |VectorArray|.
     """
     assert A.source == A.range
+    if E is None:
+        E = IdentityOperator(A.source)
```

Comment:
It should be documented that `E` is optional.

Reply:
Should we also mention that, while `r * len(b)` is an upper bound to the produced basis size, due to deflation the actual basis might be smaller?

Reply:
Did both. I also removed the 'real' everywhere from the docstring, as I don't see what it is supposed to mean exactly. The algorithm would work the same for complex data, returning an orthonormal basis for the respective $\mathbb{C}$-Krylov space.